### ISSUE TITLE
Collaboration Messages

### ISSUE THREAD

**Issue by freeflyclone** [freeflyclone/madstyletv#7]
Should be obvious what this is for. ;)

**Comment by jacarterjr** [comment]
I have pushed, creating a branched called 'carter'. It contains the changes I'm working on.

**Comment by jacarterjr** [comment]
Where should I make the changes to display the HUD attached to the "sled" you told me about?

**Comment by freeflyclone** [comment]
The GUI stuff is not as slick as I thought.  I worked all evening to fix it and didn't succeed, so it's gonna take a deep dive to sort it out.

**Comment by freeflyclone** [comment]
WorldGUIDev project shows how to create and render to XGLGuiCanvas.  The use of "shaders/gui-tex" makes it projected into world space, rather than screen space like "shaders/ortho-tex" does.  XGL has two XGLShape lists, one for world objects, one for 2D GUI objects.  The 2D GUI objects get rendered second, with GL_DEPTH testing turned off, so everything gets drawn, always.  I *think* that I want to add what is in WorldGUIDevBuildScene to the 2D list, but I'm not clear how that will work yet with regard to projection.  It'll take some experimenting.

Edit: Couldn't stand not knowing.  I've made it work as I want: added the above to the 2D list instead of the world list, activates with the '~' key, and gets rendered AFTER world objects, so it's transparent.  All that's left is to attach it to the sled. Which I'll do tonight.

**Comment by jacarterjr** [comment]
Cool. I will take a look
 Let me know when you have connectivity again.

**Comment by freeflyclone** [comment]
I have connectivity.  DSL modem got smoked, got a new one and I'm back.  OVRTest has the guiCanvas attached to the sled.  There's still some wonkiness that's gonna take some deep though to make it work exactly right, but its definitely closer.

**Comment by jacarterjr** [comment]
Okay, I'll pull and take a look.

**Comment by freeflyclone** [comment]
Fixed Layers bug.  Incredibly easy fix.
GUI layer in OVRTest is starting to look good.  Got transparency working well.  Layers only affect shapes created with AddShape(),  all child shapes thereof are rendered according to the top-level shape's layer.  THAT is gonna be tricky to fix.l

**Comment by jacarterjr** [comment]
Okay. Working on getting the physx demo working.

**Comment by freeflyclone** [comment]
In case it's not too late,  I've added PhysX build instructions to the wiki.  Just tested to verify they work ;)

**Comment by jacarterjr** [comment]
Nope not too late. Didn't get to it this weekend but will definitely do so
tomorrow.

**Comment by jacarterjr** [comment]
Alright! Just saw the text display in the OVRTest. I'll start following the PhysX build instructions tonight.
I will have to incorporate the changes from OVRTest into the CyberTank project on my branch. Unless you have other changes in mind. Perhaps I'll do it regardless just as an exercise.

**Comment by jacarterjr** [comment]
Error(s) Will Robinson!! I will attempt to resolve myself, but wanted to document in case you wanted details about the build failure.

Error	17	error LNK2019: unresolved external symbol __imp__PxCreateFoundation referenced in function "public: void __thiscall PhysXXGL::initPhysics(bool)" (?initPhysics@PhysXXGL@@QAEX_N@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	18	error LNK2019: unresolved external symbol "public: static class physx::PxProfileZoneManager & __cdecl physx::PxProfileZoneManager::createProfileZoneManager(class physx::PxFoundation *)" (?createProfileZoneManager@PxProfileZoneManager@physx@@SAAAV12@PAVPxFoundation@2@@Z) referenced in function "public: void __thiscall PhysXXGL::initPhysics(bool)" (?initPhysics@PhysXXGL@@QAEX_N@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	19	error LNK2019: unresolved external symbol __imp__PxRegisterArticulations referenced in function "class physx::PxPhysics * __cdecl PxCreatePhysics(unsigned int,class physx::PxFoundation &,class physx::PxTolerancesScale const &,bool,class physx::PxProfileZoneManager *)" (?PxCreatePhysics@@YAPAVPxPhysics@physx@@IAAVPxFoundation@2@ABVPxTolerancesScale@2@_NPAVPxProfileZoneManager@2@@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	20	error LNK2019: unresolved external symbol __imp__PxRegisterHeightFields referenced in function "class physx::PxPhysics * __cdecl PxCreatePhysics(unsigned int,class physx::PxFoundation &,class physx::PxTolerancesScale const &,bool,class physx::PxProfileZoneManager *)" (?PxCreatePhysics@@YAPAVPxPhysics@physx@@IAAVPxFoundation@2@ABVPxTolerancesScale@2@_NPAVPxProfileZoneManager@2@@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	21	error LNK2019: unresolved external symbol __imp__PxRegisterCloth referenced in function "class physx::PxPhysics * __cdecl PxCreatePhysics(unsigned int,class physx::PxFoundation &,class physx::PxTolerancesScale const &,bool,class physx::PxProfileZoneManager *)" (?PxCreatePhysics@@YAPAVPxPhysics@physx@@IAAVPxFoundation@2@ABVPxTolerancesScale@2@_NPAVPxProfileZoneManager@2@@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	22	error LNK2019: unresolved external symbol __imp__PxRegisterParticles referenced in function "class physx::PxPhysics * __cdecl PxCreatePhysics(unsigned int,class physx::PxFoundation &,class physx::PxTolerancesScale const &,bool,class physx::PxProfileZoneManager *)" (?PxCreatePhysics@@YAPAVPxPhysics@physx@@IAAVPxFoundation@2@ABVPxTolerancesScale@2@_NPAVPxProfileZoneManager@2@@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	23	error LNK2019: unresolved external symbol __imp__PxCreateBasePhysics referenced in function "class physx::PxPhysics * __cdecl PxCreatePhysics(unsigned int,class physx::PxFoundation &,class physx::PxTolerancesScale const &,bool,class physx::PxProfileZoneManager *)" (?PxCreatePhysics@@YAPAVPxPhysics@physx@@IAAVPxFoundation@2@ABVPxTolerancesScale@2@_NPAVPxProfileZoneManager@2@@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	24	error LNK2019: unresolved external symbol __imp__PxGetPhysics referenced in function "public: virtual void __thiscall PhysXXGL::PhysxRenderer::Draw(void)" (?Draw@PhysxRenderer@PhysXXGL@@UAEXXZ)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	25	error LNK2019: unresolved external symbol __imp__PxCreateCooking referenced in function "public: void __thiscall PhysXXGL::initPhysics(bool)" (?initPhysics@PhysXXGL@@QAEX_N@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	26	error LNK2019: unresolved external symbol "class physx::PxDistanceJoint * __cdecl physx::PxDistanceJointCreate(class physx::PxPhysics &,class physx::PxRigidActor *,class physx::PxTransform const &,class physx::PxRigidActor *,class physx::PxTransform const &)" (?PxDistanceJointCreate@physx@@YAPAVPxDistanceJoint@1@AAVPxPhysics@1@PAVPxRigidActor@1@ABVPxTransform@1@12@Z) referenced in function "public: void __thiscall PhysXXGL::RayCast(struct glm::tvec3<float,0>,struct glm::tvec3<float,0>)" (?RayCast@PhysXXGL@@QAEXU?$tvec3@M$0A@@glm@@0@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	27	error LNK2019: unresolved external symbol "class physx::PxSphericalJoint * __cdecl physx::PxSphericalJointCreate(class physx::PxPhysics &,class physx::PxRigidActor *,class physx::PxTransform const &,class physx::PxRigidActor *,class physx::PxTransform const &)" (?PxSphericalJointCreate@physx@@YAPAVPxSphericalJoint@1@AAVPxPhysics@1@PAVPxRigidActor@1@ABVPxTransform@1@12@Z) referenced in function "public: class physx::PxSphericalJoint * __thiscall <lambda_40b1eb971e2bb3254c2546ac0799350b>::operator()(class physx::PxRigidActor *,class physx::PxTransform const &,class physx::PxRigidActor *,class physx::PxTransform const &)const " (??R<lambda_40b1eb971e2bb3254c2546ac0799350b>@@QBEPAVPxSphericalJoint@physx@@PAVPxRigidActor@2@ABVPxTransform@2@01@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	28	error LNK2019: unresolved external symbol "class physx::PxFlags<enum physx::PxFilterFlag::Enum,unsigned short> __cdecl physx::PxDefaultSimulationFilterShader(unsigned int,struct physx::PxFilterData,unsigned int,struct physx::PxFilterData,class physx::PxFlags<enum physx::PxPairFlag::Enum,unsigned short> &,void const *,unsigned int)" (?PxDefaultSimulationFilterShader@physx@@YA?AV?$PxFlags@W4Enum@PxFilterFlag@physx@@G@1@IUPxFilterData@1@I0AAV?$PxFlags@W4Enum@PxPairFlag@physx@@G@1@PBXI@Z) referenced in function "public: void __thiscall PhysXXGL::initPhysics(bool)" (?initPhysics@PhysXXGL@@QAEX_N@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	29	error LNK2019: unresolved external symbol "public: __thiscall physx::PxDefaultErrorCallback::PxDefaultErrorCallback(void)" (??0PxDefaultErrorCallback@physx@@QAE@XZ) referenced in function "public: __thiscall PhysXXGL::PhysXXGL(void)" (??0PhysXXGL@@QAE@XZ)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	30	error LNK2019: unresolved external symbol "public: virtual __thiscall physx::PxDefaultErrorCallback::~PxDefaultErrorCallback(void)" (??1PxDefaultErrorCallback@physx@@UAE@XZ) referenced in function __unwindfunclet$??0PhysXXGL@@QAE@XZ$3	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	31	error LNK2019: unresolved external symbol "public: static bool __cdecl physx::PxRigidBodyExt::updateMassAndInertia(class physx::PxRigidBody &,float,class physx::PxVec3 const *,bool)" (?updateMassAndInertia@PxRigidBodyExt@physx@@SA_NAAVPxRigidBody@2@MPBVPxVec3@2@_N@Z) referenced in function "public: void __thiscall PhysXXGL::createStack(class physx::PxTransform const &,unsigned int,float)" (?createStack@PhysXXGL@@QAEXABVPxTransform@physx@@IM@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	32	error LNK2019: unresolved external symbol "class physx::PxDefaultCpuDispatcher * __cdecl physx::PxDefaultCpuDispatcherCreate(unsigned int,unsigned int *)" (?PxDefaultCpuDispatcherCreate@physx@@YAPAVPxDefaultCpuDispatcher@1@IPAI@Z) referenced in function "public: void __thiscall PhysXXGL::initPhysics(bool)" (?initPhysics@PhysXXGL@@QAEX_N@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	33	error LNK2019: unresolved external symbol "class physx::PxRigidDynamic * __cdecl physx::PxCreateDynamic(class physx::PxPhysics &,class physx::PxTransform const &,class physx::PxGeometry const &,class physx::PxMaterial &,float,class physx::PxTransform const &)" (?PxCreateDynamic@physx@@YAPAVPxRigidDynamic@1@AAVPxPhysics@1@ABVPxTransform@1@ABVPxGeometry@1@AAVPxMaterial@1@M1@Z) referenced in function "public: class physx::PxRigidDynamic * __thiscall PhysXXGL::createDynamic(class physx::PxTransform const &,class physx::PxGeometry const &,class physx::PxMaterial &,float)" (?createDynamic@PhysXXGL@@QAEPAVPxRigidDynamic@physx@@ABVPxTransform@3@ABVPxGeometry@3@AAVPxMaterial@3@M@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	34	error LNK2019: unresolved external symbol "class physx::PxRigidStatic * __cdecl physx::PxCreatePlane(class physx::PxPhysics &,class physx::PxPlane const &,class physx::PxMaterial &)" (?PxCreatePlane@physx@@YAPAVPxRigidStatic@1@AAVPxPhysics@1@ABVPxPlane@1@AAVPxMaterial@1@@Z) referenced in function "public: void __thiscall PhysXXGL::initPhysics(bool)" (?initPhysics@PhysXXGL@@QAEX_N@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	35	error LNK2019: unresolved external symbol _PxInitExtensions referenced in function "public: void __thiscall PhysXXGL::initPhysics(bool)" (?initPhysics@PhysXXGL@@QAEX_N@Z)	D:\code\madstyletv\examples\PhysXTest\physx-xgl.obj	PhysXTest
Error	36	error LNK1120: 19 unresolved externals	D:\code\madstyletv\Debug\PhysXTest.exe	PhysXTest

**Comment by freeflyclone** [comment]
did you set PHYSX_DIR environment variable?

**Comment by jacarterjr** [comment]
Yes. Followed the instructions to the letter. Libraries are being picked
up. Viewed symbols in physx libs using dumpbin and found the exported
functions. However there prefix __imp_ * isn't there.

**Comment by freeflyclone** [comment]
Hmm.  I've seen  that before...  I'll get back to you.

**Comment by freeflyclone** [comment]
That looks like static vs debug linking screw up... I **might** have changed the PhysX default runtime (DLL vs static).

**Comment by freeflyclone** [comment]
Shure as shit.  That's exactly what I did.

**Comment by freeflyclone** [comment]
Here's a git patch from my PhysX-3.3 git repo:

[0001-Switched-all-projects-in-vc12win64-over-to-MultiThre.zip](https://github.com/freeflyclone/madstyletv/files/1326993/0001-Switched-all-projects-in-vc12win64-over-to-MultiThre.zip)

Had to zip it for github to take it.  Extract and apply the patch 

https://stackoverflow.com/questions/6658313/generate-a-git-patch-for-a-specific-commit.  I ran it from git-bash shell in the root of the PhysX-3.3 dir

**Comment by jacarterjr** [comment]
Okay. Performed the patch. And the same thing...

**Comment by freeflyclone** [comment]
You're right, my patch doesn't solve that problem.  I just tried it.  It DOES solve another problem which you would have run into, but VS 2013 accurately reports the actual problem in that case:

PhysX3ExtensionsDEBUG.lib(ExtDistanceJoint.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in ExampleGUI.obj
...

Will keep looking for your issue.

**Comment by freeflyclone** [comment]
I reproduced your error stream:  I switched the build Platform to Win32 in madstyletv, then rebuilt the solution.  So... check your build platform setting:  here's what I have:

![solution-platform](https://user-images.githubusercontent.com/8210646/30784914-a52d0ff4-a112-11e7-9cb9-28f27a9f798f.png)

Note that this is **not** the default platform when you first open the solution. 

The patch I supplied only patches the x64 flavor of PhysXSDK.  I suppose you could just as easily patch the 32-bit version.  Pretty sure I did some bit of cleverness to batch edit the .vcxproj files in PhysXSDK to change "Multithreaded" to "MultithreadedDLL" .

**Comment by freeflyclone** [comment]
Updated PhysXSDK build instructions in the wiki.

**Comment by freeflyclone** [comment]
Flight controls working!  You were right: forward vector * orientation added to current position.  OVRTest project demonstrates.

**Comment by jacarterjr** [comment]
Awesome! I'll check it out.

**Comment by freeflyclone** [comment]
HMD sled now working the way I intended.  That is: it's the cockpit and flies through the world, with local things like VR hand objects and HUD GUI objects attached to it.  The key was taking the inverse model matrix for use with the HMD view matrix.  Also, lighting now works propertly, specular reflection follows the sled and head movement.

**Comment by freeflyclone** [comment]
Got the PhysX demo working with the sled.  And it's HORRIBLY slow.  Not sure where the bottleneck is yet.  Not even sure how to go about diagnosing the problem.  Suspicious of PhysX CPU bounding as I've observed it problematic even with out OVR overhead.

Will probably hook up the Touch controllers just for grins then commit/push it.

**Comment by freeflyclone** [comment]
Considering a switch to Bullet physics.

**Comment by freeflyclone** [comment]
Sidetracked by ComputeShaderTest bug I introduced awhile back without knowing it.  Fixed now.
Bullet physics is POORLY documented.

**Comment by freeflyclone** [comment]
News flash: PhysX is slow... in Debug configuration.  Seems fine in Release configuration.  

However, I had to tweek Physx to get it to run properly with mesh based collision objects.  Need to figure out how to generate another git patch file so you can modify your copy accordingly.

**Comment by freeflyclone** [comment]
Update: Added new patch file, changed wiki PhysX Build page to point to it.  It's checked into the *patches* subfolder.  Tested briefly: still can't do as many PhysX objects as without the HMD, but it's WAY better than debug mode.

**Comment by jacarterjr** [comment]
I'll compile and speed test after the push.

**Comment by freeflyclone** [comment]
I pushed right before my last post.  So it's there.
